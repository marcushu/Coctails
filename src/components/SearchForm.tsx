import { Box, Button, TextField, Typography } from "@mui/material";
import { FunctionComponent, useState } from "react";

interface SearchFormProps {
  callback: (tearm: string, byName: boolean) => void
}

const SearchForm: FunctionComponent<SearchFormProps> = ({ callback }) => {
  const [searchByNameTxt, setSearchByNameTxt] = useState("");
  const [ingredientTxt, setingredientTxt] = useState("");

  const callSearch = () => {
    if (!searchByNameTxt && !ingredientTxt) {
      alert("Fill in a name or an ingredeint")
    } else if (searchByNameTxt && ingredientTxt) {
      alert('Fill in EITHER a coctail name OR an ingredient')
    } else {
      callback(searchByNameTxt! + ingredientTxt, !!searchByNameTxt)
    }
  }

  return (
    <Box
      margin='auto'
      display='flex'
      flexDirection={['column', 'row']}
      alignItems='center'
      justifyContent='space-between'
      maxWidth='640px'>
      <TextField
        variant="standard"
        id="coctailName"
        placeholder="coctail name"
        size="small"
        fullWidth={true}
        value={searchByNameTxt}
        onMouseDown={() => setingredientTxt("")}
        onChange={e => setSearchByNameTxt(e.target.value)}
        sx={{
          input: { color: 'white' }, maxWidth: ['1000', '210px'],
          borderBottom: '1px solid white'
        }}
        data-testid="coctailname" />
      <Typography color='white' fontWeight='bold' pt={[3, 'unset']}> ~ or ~ </Typography>
      <TextField
        variant="standard"
        id="ingredient"
        placeholder="an ingredient"
        size="small"
        fullWidth={true}
        value={ingredientTxt}
        onMouseDown={() => setSearchByNameTxt("")}
        onChange={e => setingredientTxt(e.target.value)}
        sx={{
          input: { color: 'white' }, maxWidth: ['1000px', '210px'],
          borderBottom: '1px solid white'
        }}
        data-testid="coctailingredient" />
      <Button
        onClick={callSearch}
        fullWidth={true}
        variant='contained'
        sx={{
          height: '40px', backgroundColor: '#b69c71',
          color: 'black', maxWidth: ['1000px', '80px'], marginTop: ['20px', '0px']
        }}>
        find
      </Button>
    </Box>
  );
}

export default SearchForm;