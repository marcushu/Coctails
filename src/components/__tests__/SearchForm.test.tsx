import { fireEvent, render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event";
import SearchForm from "../SearchForm"

const mockFunc = jest.fn();

describe('SearchFrom', () => {

    it("renders", () => {
        const { asFragment } = render(<SearchForm callback={mockFunc} />);

        expect(asFragment).toMatchSnapshot();
    });

    it("calls the callback with correct coctail name parameters", () => {
        render(<SearchForm callback={mockFunc} />);

        const findBtn = screen.getByRole('button', { name: /find/i });
        const nameInput = screen.getByTestId('coctailname');

        userEvent.type(nameInput, 'martini');
        fireEvent.click(findBtn);

        expect(mockFunc).toHaveBeenCalled();
        expect(mockFunc.mock.calls[0][0]).toBe('martini');
        expect(mockFunc.mock.calls[0][1]).toBe(true);

    });

    it('calls the callback with correct coctail ingredient parameters', () => {
        render(<SearchForm callback={mockFunc} />);

        const findBtn = screen.getByRole('button', { name: /find/i });
        const ingredient = screen.getByTestId('coctailingredient');

        userEvent.type(ingredient, 'vermouth');
        fireEvent.click(findBtn);

        expect(mockFunc).toHaveBeenCalled();
        expect(mockFunc.mock.calls[0][0]).toBe('vermouth');
        expect(mockFunc.mock.calls[0][1]).toBe(false);
    });

    it('will not call the callback without input values', () => {
        render(<SearchForm callback={mockFunc} />);

        const findBtn = screen.getByRole('button', { name: /find/i });

        fireEvent.click(findBtn);

        expect(mockFunc).not.toHaveBeenCalled();
    });

    it('will not call callback with both fields filled', () => {
        render(<SearchForm callback={mockFunc} />);

        const findBtn = screen.getByRole('button', { name: /find/i });
        const ingredient = screen.getByTestId('coctailingredient');
        const nameInput = screen.getByTestId('coctailname');

        userEvent.type(ingredient, 'vermouth');
        userEvent.type(nameInput, 'martini');
        fireEvent.click(findBtn);

        expect(mockFunc).not.toHaveBeenCalled();
    })
})