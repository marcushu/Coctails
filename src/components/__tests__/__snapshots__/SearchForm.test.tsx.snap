// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchFrom renders 1`] = `[Function]`;
